File: src/jfan/fan/sys/StrInStream.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   1 Aug 06  Brian Frank  Creation
//
package fan.sys;

import java.io.*;

/**
 * StrInStream
 */
public class StrInStream
  extends InStream
{

//////////////////////////////////////////////////////////////////////////
// Constructor
//////////////////////////////////////////////////////////////////////////

  public StrInStream(String val)
  {
    this.str  = val;
    this.size = val.length();
    this.pos  = 0;
  }

//////////////////////////////////////////////////////////////////////////
// InStream
//////////////////////////////////////////////////////////////////////////

  public int r()
  {
    return rChar();
  }

  public Long read()
  {
    int b = r(); return (b < 0) ? null : FanInt.pos[b & 0xFF];
  }

  public Long readBuf(Buf buf, Long n)
  {
    int nval = n.intValue();
    for (int i=0; i<nval; ++i)
    {
      int c = rChar();
      if (c < 0) return Long.valueOf(i);
      buf.out.w(c);
    }
    return n;
  }

  public InStream unread(Long c)
  {
    return unreadChar(c);
  }

  public int rChar()
  {
    if (pushback != null && pushback.sz() > 0)
      return ((Long)pushback.pop()).intValue();
    if (pos >= size) return -1;
    return str.charAt(pos++);
  }

  public Long readChar()
  {
    if (pushback != null && pushback.sz() > 0)
      return (Long)pushback.pop();
    if (pos >= size) return null;
    return Long.valueOf(str.charAt(pos++));
  }

  public InStream unreadChar(Long c)
  {
    if (pushback == null) pushback = new List(Sys.IntType, 8);
    pushback.push(c);
    return this;
  }

  public boolean close()
  {
    return true;
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  String str;
  int pos;
  int size;
  List pushback;

}

File: src/nfan/fan/sys/StrInStream.cs
//
// Copyright (c) 2007, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   12 Apr 07  Andy Frank  Creation
//

namespace Fan.Sys
{
  /// <summary>
  /// StrInStream.
  /// </summary>
  public class StrInStream : SysInStream
  {

  //////////////////////////////////////////////////////////////////////////
  // Constructor
  //////////////////////////////////////////////////////////////////////////

    public StrInStream(string val) : base(null)
    {
      m_str  = val;
      m_size = val.Length;
      m_pos  = 0;
    }

  //////////////////////////////////////////////////////////////////////////
  // InStream
  //////////////////////////////////////////////////////////////////////////

    public override int r()
    {
      return rChar();
    }

    public override Long read()
    {
      int b = r(); return (b < 0) ? null : Long.valueOf(b & 0xFF);
    }

    public override Long readBuf(Buf buf, long n)
    {
      int nval = (int)n;
      for (int i=0; i<nval; ++i)
      {
        int c = rChar();
        if (c < 0) return Long.valueOf(i);
        buf.m_out.w(c);
      }
      return Long.valueOf(n);
    }

    public override InStream unread(long c)
    {
      return unreadChar(c);
    }

    public override int rChar()
    {
      if (m_pushback != null && m_pushback.sz() > 0)
        return ((Long)m_pushback.pop()).intValue();
      if (m_pos >= m_size) return -1;
      return m_str[m_pos++];
    }

    public override Long readChar()
    {
      if (m_pushback != null && m_pushback.sz() > 0)
        return (Long)m_pushback.pop();
      if (m_pos >= m_size) return null;
      return Long.valueOf(m_str[m_pos++]);
    }

    public override InStream unreadChar(long c)
    {
      if (m_pushback == null) m_pushback = new List(Sys.IntType, 8);
      m_pushback.push(Long.valueOf(c));
      return this;
    }

    public override bool close()
    {
      return true;
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    internal string m_str;
    internal int m_pos;
    internal int m_size;
    internal List m_pushback;

  }

}